jsk_pcl_ros/src/moving_least_square_smoothing_nodelet.cpp

```diff
@@ -0,0 +1,121 @@
+ *  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ *  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ *********************************************************************/
+
+#define BOOST_PARAMETER_MAX_ARITY 7
+#include "jsk_pcl_ros/moving_least_square_smoothing.h"
+
+#include <geometry_msgs/PoseStamped.h>
+#include <geometry_msgs/PointStamped.h>
+
+#include "jsk_pcl_ros/pcl_conversion_util.h"
+
+namespace jsk_pcl_ros
+{
+  void MovingLeastSquareSmoothing::smooth(const sensor_msgs::PointCloud2ConstPtr& input)
+  {
+    boost::mutex::scoped_lock lock(mtx_);
+    pcl::PointCloud<pcl::PointXYZRGB>::Ptr cloud(new pcl::PointCloud<pcl::PointXYZRGB>);
+    pcl::PointCloud<pcl::PointXYZRGB>::Ptr result_cloud(new pcl::PointCloud<pcl::PointXYZRGB>);
+    pcl::fromROSMsg(*input, *cloud);
+
+    std::vector<int> indices;
+    cloud->is_dense = false;
+    pcl::removeNaNFromPointCloud(*cloud, *cloud, indices);
+
+    smoother_.setInputCloud (cloud);
+    smoother_.process (*result_cloud);
+
+    sensor_msgs::PointCloud2 pointcloud2;
+    pcl::toROSMsg(*result_cloud, pointcloud2);
+    pointcloud2.header.frame_id = input->header.frame_id;
+    pointcloud2.header.stamp = input->header.stamp;
+    pub_.publish(pointcloud2);
+  }
+  
+  void MovingLeastSquareSmoothing::subscribe()
+  {
+    sub_input_ = pnh_->subscribe("input", 1, &MovingLeastSquareSmoothing::smooth, this);
+  }
+
+  void MovingLeastSquareSmoothing::unsubscribe()
+  {
+    sub_input_.shutdown();
+  }
+
+  void MovingLeastSquareSmoothing::config_callback(Config &config, uint32_t level)
+  {
+    boost::mutex::scoped_lock lock(mtx_);
+    search_radius_ = config.search_radius;
+    gauss_param_set_ = config.gauss_param_set;
+    use_polynomial_fit_ = config.use_polynomial_fit;
+    polynomial_order_ = config.polynomial_order;
+    calc_normal_ = config.calc_normal;
+
+    smoother_.setComputeNormals (calc_normal_);
+    smoother_.setSearchRadius (search_radius_);
+    if (gauss_param_set_) smoother_.setSqrGaussParam (gauss_param_set_);
+    smoother_.setPolynomialFit (use_polynomial_fit_);
+    smoother_.setPolynomialOrder (polynomial_order_);
+    typename pcl::search::KdTree<pcl::PointXYZRGB>::Ptr tree (new typename pcl::search::KdTree<pcl::PointXYZRGB> ());
+    smoother_.setSearchMethod (tree);
+  }
+
+  void MovingLeastSquareSmoothing::onInit(void)
+  {
+    DiagnosticNodelet::onInit();
+    pnh_->param("gauss_param_set", gauss_param_set_, false);
+    pnh_->param("search_radius", search_radius_, 0.03);
+    pnh_->param("use_polynomial_fit", use_polynomial_fit_, false);
+    pnh_->param("polynomial_order", polynomial_order_, 2);
+    pnh_->param("calc_normal", calc_normal_, true);
+    pub_ =advertise<sensor_msgs::PointCloud2>(*pnh_, "output", 1);
```

Comment:
publisher should be instantiated at the end of onInit.
parameter should be initialized before instantiating publisher.